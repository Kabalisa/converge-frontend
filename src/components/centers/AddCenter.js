/* eslint-disable import/no-named-as-default,array-callback-return,consistent-return */
import React, { Component } from 'react';
import { compose, graphql } from 'react-apollo';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import { Input } from '../commons';
import ActionButtons from '../commons/ActionButtons';
import ADD_CENTER_MUTATION from '../../graphql/mutations/centers';
import MrmModal from '../commons/Modal';
import countries from '../../fixtures/countries';
import notification from '../../utils/notification';
import { GET_USER_QUERY } from '../../graphql/queries/People';
import { decodeTokenAndGetUserData } from '../../utils/Cookie';
import allCites from '../../fixtures/cities';

export class AddCenter extends Component {
  static propTypes = {
    addCenter: PropTypes.func.isRequired,
  };

  state = {
    centerName: '',
    abbreviation: '',
    closeModal: false,
  };

  /**
   * it closes the modal and resets imageUrl and thumbnailName to defaults
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({
      closeModal: true,
    });
  };

  /**
   * updates the state value of a given input
   * @param event
   * @returns {void}
   */
  handleInputChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  /**
   * updates closeModal state to false when ever modal closes
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   *1. submits center data to the backend api
   *2. notifies the user about the response from the request
   *
   * @returns {void}
   */
  createCenter = () => {
    const {
      centerName, abbreviation,
    } = this.state;

    const { user } = this.props;
    const { location } = user.user;
    let selectedCountry = {};

    allCites.find((data) => {
      if (data.city.toLowerCase() === location.toLowerCase()) {
        selectedCountry = countries.find(singleCountry => (singleCountry.id === data.country));
      }
    });

    const { timeZone, name } = selectedCountry;
    this.props
      .addCenter({
        variables: {
          country: name,
          name: centerName,
          timeZone,
          abbreviation,
        },
      })
      .then(() => {
        notification(
          toastr,
          'success',
          `${centerName} center has been added successfully`,
        )();
        this.props.refetch();
      })
      .catch((err) => {
        notification(toastr, 'error', err.graphQLErrors[0].message)();
      });
    this.handleCloseModal();
  };

  /**
   *1. validates input
   *2. calls createCenter method after validation to perform center creation
   *
   * @param {object} event
   * @returns {void}
   */
  handleAddCenter = (event) => {
    event.preventDefault();
    const {
      centerName, abbreviation,
    } = this.state;
    if (!centerName) {
      notification(toastr, 'error', 'center name is required')();
    } else if (!abbreviation) {
      notification(toastr, 'error', 'abbreviation field is required')();
    } else {
      this.createCenter();
    }
  };

  render() {
    const {
      centerName, closeModal,
      abbreviation,
    } = this.state;
    return (
      <MrmModal
        title="ADD CENTER"
        buttonText="Add Center"
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
        className="add-office-modal"
        modalButton="add-button"
      >
        <form className="modal-form">
          <Input
            labelName="Center Name"
            name="centerName"
            value={centerName}
            placeholder="Enter center name"
            id="centerName"
            onChange={this.handleInputChange}
          />
          <Input
            labelName="Abbreviation"
            name="abbreviation"
            value={abbreviation}
            placeholder="Eg. Kla"
            id="abbreviation"
            onChange={this.handleInputChange}
          />
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            actionButtonText="ADD CENTER"
            onClickSubmit={this.handleAddCenter}
          />
        </form>
      </MrmModal>
    );
  }
}

AddCenter.propTypes = {
  refetch: PropTypes.func.isRequired,
  user: PropTypes.shape({
    user: PropTypes.object,
  }).isRequired,
};

const { UserInfo: userData } = decodeTokenAndGetUserData() || {};

export default compose(
  graphql(ADD_CENTER_MUTATION, { name: 'addCenter' }),
  graphql(GET_USER_QUERY, {
    name: 'user',
    options: /* istanbul ignore next */ () => ({
      variables: {
        email:
          process.env.NODE_ENV === 'test'
            ? 'sammy.muriuki@andela.com'
            : userData.email,
      },
    }),
  }),
)(AddCenter);
