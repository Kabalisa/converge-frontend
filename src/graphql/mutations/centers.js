import gql from 'graphql-tag';

const ADD_CENTER_MUTATION = gql`
  mutation createLocation($imageUrl: String, $name: String!, $timeZone:String!, $country: String!, $abbreviation: String! ) {
    createLocation(imageUrl: $imageUrl, name: $name, timeZone: $timeZone, country: $country, abbreviation: $abbreviation) {
      location {
        id
        name
      }
    }
  }
`;

const EDIT_CENTER_MUTATION = gql`
  mutation updateLocation($locationId: Int!, $name: String!, $country: String!, $abbreviation: String!){
    updateLocation(locationId: $locationId, name: $name, country: $country, abbreviation: $abbreviation){ 
        location{
            name
            abbreviation
            country
        }
    }
}
`;

const DELETE_CENTER_MUTATION = gql`
  mutation deleteLocation($locationId: Int!){
    deleteLocation(locationId:$locationId){
        location{
          id
        }
    }
  }
`;
export {
  ADD_CENTER_MUTATION as default,
  EDIT_CENTER_MUTATION,
  DELETE_CENTER_MUTATION,
};
