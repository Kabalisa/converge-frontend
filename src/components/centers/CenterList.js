import React from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import MenuTitle from '../commons/MenuTitle';
import AddCenter from './AddCenter'; // eslint-disable-line
import { GET_ALL_CENTERS } from '../../graphql/queries/centers';
import ColGroup from '../helpers/ColGroup';
import TableHead from '../helpers/TableHead';
import Location from './Center';
import Spinner from '../commons/Spinner';
import { GET_USER_ROLE } from '../../graphql/queries/People';
import { decodeTokenAndGetUserData } from '../../utils/Cookie';
import { saveItemInLocalStorage } from '../../utils/Utilities';
import DataNotFound from '../commons/DataNotFound';
import Errors from '../commons/Errors';

export const CenterList = (props) => {
  const {
    loading, refetch, allLocations, error,
  } = props.data;
  const { user } = props.user;

  if (loading) {
    return <Spinner />;
  }
  if (user && user.roles) saveItemInLocalStorage('access', user.roles[0].id);
  return (
    <div className="settings-locations">
      <div className="settings-locations-control">
        <MenuTitle title="Centers" />
        <AddCenter refetch={refetch} />
      </div>
      <div className="settings-locations-list">
        {
        !error ?
        (
          allLocations.length < 1
          ? <DataNotFound /> :
          <table>
            <ColGroup />
            <TableHead titles={['Center', 'Country', 'Abbreviation', 'Action']} />
            <tbody>
              {allLocations && allLocations.map(location => (
                <Location
                  location={location}
                  key={location.id}
                  refetch={refetch}
                />
             ))}
            </tbody>
          </table>
        ) :
          <Errors message="Data cannot be returned at the moment" />
      }
      </div>
    </div>
  );
};

CenterList.propTypes = {
  data: PropTypes.shape({
    loading: PropTypes.bool,
    error: PropTypes.object,
    refetch: PropTypes.func,
    allLocations: PropTypes.array,
  }),
  user: PropTypes.shape({
    user: PropTypes.object,
  }).isRequired,
};

CenterList.defaultProps = {
  data: {
    loading: true,
  },
};

const { UserInfo: userData } = decodeTokenAndGetUserData() || {};

export default compose(
  graphql(GET_ALL_CENTERS),
  graphql(GET_USER_ROLE, {
    name: 'user',
    options: /* istanbul ignore next */ () => ({
      variables: {
        email:
          process.env.NODE_ENV === 'test'
            ? 'sammy.muriuki@andela.com'
            : userData.email,
      },
    }),
  }),
)(CenterList);
