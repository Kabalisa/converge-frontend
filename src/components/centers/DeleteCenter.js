import React from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import toastr from 'toastr';
import ActionButtons from '../commons/ActionButtons';
import MrmModal from '../commons/Modal';
import { DELETE_CENTER_MUTATION } from '../../graphql/mutations/centers';
import notification from '../../utils/notification';

import '../../assets/styles/deleteModal.scss';

/**
 * Delete center Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class DeleteCenter extends React.Component {
  static propTypes = {
    centerName: PropTypes.string.isRequired,
    centerId: PropTypes.string.isRequired,
    deleteCenter: PropTypes.func.isRequired,
    refetch: PropTypes.func,
  };

  state = {
    closeModal: false,
    isDeleting: false,
  };

  /**
   * It closes a modal
   *
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({ closeModal: true });
  };

  /**
   * Handles the state changes for the deleting location modal
   *
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   * This function deletes a location
   *
   * @returns {void}
   */
  handleDeleteCenter = () => {
    const {
      centerId, deleteCenter, centerName, refetch,
    } = this.props;
    this.toggleLoading();
    deleteCenter({
      variables: {
        locationId: centerId,
      },
    })
      .then(() => {
        this.toggleLoading();
        this.handleCloseModal();
        notification(toastr, 'success', `${centerName} has been deleted successfully`)();
        refetch();
      })
      .catch((err) => {
        this.toggleLoading();
        this.handleCloseModal();
        notification(toastr, 'error', err.graphQLErrors[0].message)();
      });
  };

  /**
   * 1. change isLoading state to it's opposite value
   * i.e true to false or vise verser
   *
   * @returns {void}
   */
  toggleLoading = () => {
    this.setState({
      isDeleting: !this.state.isDeleting,
    });
  }

  render() {
    const { closeModal, isDeleting } = this.state;

    return (
      <MrmModal
        className="delete-modal"
        title="DELETE CENTER"
        buttonText="Delete"
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
      >
        <div className="delete-modal-content">
          <p id="confirm-msg">
            Are you sure you want to delete the {`"${this.props.centerName}"`}{' '}
            center? This cannot be undone & all data will be lost
          </p>
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            isLoading={isDeleting}
            actionButtonText="DELETE CENTER"
            onClickSubmit={this.handleDeleteCenter}
          />
        </div>
      </MrmModal>
    );
  }
}

DeleteCenter.defaultProps = {
  refetch: () => {},
};

export default compose(
  graphql(DELETE_CENTER_MUTATION, { name: 'deleteCenter' }),
)(DeleteCenter);
