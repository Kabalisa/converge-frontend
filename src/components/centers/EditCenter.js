/* eslint-disable import/no-named-as-default,array-callback-return,consistent-return */
import React, { Component } from 'react';
import { compose, graphql } from 'react-apollo';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import { Input } from '../commons';
import ActionButtons from '../commons/ActionButtons';
import { EDIT_CENTER_MUTATION } from '../../graphql/mutations/centers';
import MrmModal from '../commons/Modal';
import countries from '../../fixtures/countries';
import notification from '../../utils/notification';
import { GET_USER_QUERY } from '../../graphql/queries/People';
import { decodeTokenAndGetUserData } from '../../utils/Cookie';
import allCites from '../../fixtures/cities';

export class EditCenter extends Component {
  state = {
    centerName: this.props.centerName,
    abbreviation: this.props.abbreviation,
    centerId: this.props.centerId,
    closeModal: false,
    isLoading: false,
  };

  /**
   * it closes the modal
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({
      closeModal: true,
      isLoading: false,
    });
  };

  /**
   * updates the state value of a given input
   * @param event
   * @returns {void}
   */
  handleInputChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  /**
   * updates closeModal state to false when ever modal closes
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   *1. submits location data to the backend api
   *2. notifies the user about the response from the request
   *
   * @returns {void}
   */
  editCenter = () => {
    const {
      centerName, abbreviation, centerId,
    } = this.state;

    const { user } = this.props;
    const { location } = user.user;
    let selectedCountry = {};

    allCites.find((data) => {
      if (data.city.toLowerCase() === location.toLowerCase()) {
        selectedCountry = countries.find(singleCountry => (singleCountry.id === data.country));
      }
    });

    const { name } = selectedCountry;
    const { refetch } = this.props;
    this.toggleLoading();
    this.props
      .editCenter({
        variables: {
          locationId: centerId,
          name: centerName,
          country: name,
          abbreviation,
        },
      })
      .then(() => {
        this.toggleLoading();
        notification(
          toastr,
          'success',
          `${centerName} location has been edited successfully`,
        )();
        refetch();
        this.handleCloseModal();
      })
      .catch((err) => {
        this.setState({ centerName: this.props.centerName });
        this.toggleLoading();
        this.handleCloseModal();
        notification(toastr, 'error', err.graphQLErrors[0].message)();
      });
    this.toggleLoading();
  };

  /**
   * 1. change isLoading state to it's opposite value
   * i.e true to false or vise verser
   *
   * @returns {void}
   */
  toggleLoading = () => {
    this.setState({
      isLoading: !this.state.isLoading,
    });
  }
  /**
   *1. validates input
   *2. calls editCenter method after validation to perform location creation
   *
   * @param {object} event
   * @returns {void}
   */
  handleEditCenter = (event) => {
    event.preventDefault();
    const {
      centerName, abbreviation,
    } = this.state;
    if (!centerName) {
      notification(toastr, 'error', 'location name is required')();
    } else if (!abbreviation) {
      notification(toastr, 'error', 'abbreviation field is required')();
    } else {
      this.toggleLoading();
      this.editCenter();
    }
  };

  render() {
    const {
      centerName, closeModal,
      abbreviation, isLoading,
    } = this.state;

    return (
      <MrmModal
        title="EDIT CENTER"
        buttonText="Edit"
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
        className="add-office-modal"
        modalButtonClassName="edit-button"
      >
        <form className="modal-form">
          <Input
            labelName="Center Name"
            name="centerName"
            value={centerName}
            placeholder="Enter center name"
            id="centerName"
            onChange={this.handleInputChange}
          />
          <Input
            labelName="Abbreviation"
            name="abbreviation"
            value={abbreviation}
            placeholder="Eg. Kla"
            id="abbreviation"
            onChange={this.handleInputChange}
          />
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            actionButtonText="SAVE CHANGES"
            isLoading={isLoading}
            onClickSubmit={this.handleEditCenter}
          />
        </form>
      </MrmModal>
    );
  }
}
EditCenter.propTypes = {
  editCenter: PropTypes.func,
  centerName: PropTypes.string,
  centerId: PropTypes.string,
  abbreviation: PropTypes.string,
  user: PropTypes.shape({
    user: PropTypes.object,
  }).isRequired,
  refetch: PropTypes.func,
};

EditCenter.defaultProps = {
  refetch: null,
  editCenter: null,
  centerName: '',
  abbreviation: '',
  centerId: null,
};

const { UserInfo: userData } = decodeTokenAndGetUserData() || {};

export default compose(
  graphql(EDIT_CENTER_MUTATION, { name: 'editCenter' }),
  graphql(GET_USER_QUERY, {
    name: 'user',
    options: /* istanbul ignore next */ () => ({
      variables: {
        email:
          process.env.NODE_ENV === 'test'
            ? 'sammy.muriuki@andela.com'
            : userData.email,
      },
    }),
  }),
)(EditCenter);
