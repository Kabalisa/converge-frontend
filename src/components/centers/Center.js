import React from 'react';
import PropTypes from 'prop-types';
import EditCenter from './EditCenter'; //eslint-disable-line
import DeleteCenter from './DeleteCenter'; //eslint-disable-line

/**
 * @param {object} props
 * @returns {JSX}
 */
const Center = (props) => {
  const {
    name, country, id, abbreviation,
  } = props.location;
  const { refetch } = props;
  return (
    <tr>
      <td>{name}</td>
      <td>{country.split('.')[1]}</td>
      <td>{abbreviation}</td>
      <td>
        <EditCenter
          id="edit-modal"
          centerName={name}
          country={country}
          abbreviation={abbreviation}
          centerId={id}
          refetch={refetch}
        />
      &nbsp;
        <DeleteCenter
          centerName={name}
          centerId={id}
          refetch={refetch}
          id="delete-modal"
        />
      </td>
    </tr>
  );
};

Center.propTypes = {
  location: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    country: PropTypes.string.isRequired,
  }).isRequired,
  refetch: PropTypes.func.isRequired,
};

export default Center;
