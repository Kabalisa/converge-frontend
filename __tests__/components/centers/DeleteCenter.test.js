import React from 'react';
import { shallow } from 'enzyme';
import { DeleteCenter } from '../../../src/components/centers/DeleteCenter';

describe('Edit center component', () => {
  const refetch = jest.fn();
  const deleteCenter = jest.fn();
  const props = {
    centerId: '1',
    centerName: 'Lagos',
    refetch,
    deleteCenter,
    isDeleting: false,
  };
  const wrapper = shallow(<DeleteCenter {...props} />);
  const deletedCenter = { centerName: 'Lagos', CenterId: 1 };
  const result = { data: { deletedCenter } };
  it('renders without failing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set closeModal state to false', () => {
    wrapper.setState({ closeModal: true });
    wrapper.instance().handleModalStateChange();
    expect(wrapper.state('closeModal')).toEqual(false);
  });

  it('should close modal', () => {
    wrapper.instance().handleCloseModal();
    expect(wrapper.state('closeModal')).toEqual(true);
  });

  it('should delete center succesfully', () => {
    const newProps = {
      deleteCenter: jest.fn(() => Promise.resolve(result)),
    };
    const newWrapper = shallow(<DeleteCenter centerName="Lagos" centerId="1" {...newProps} />);

    newWrapper.setState({ closeModal: false });
    newWrapper.instance().handleDeleteCenter();
    expect(newProps.deleteCenter).toHaveBeenCalled();
  });
});
