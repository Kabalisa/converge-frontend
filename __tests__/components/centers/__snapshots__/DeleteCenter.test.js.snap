// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit center component renders without failing 1`] = `
<MrmModal
  buttonText="Delete"
  className="delete-modal"
  closeModal={false}
  handleCloseRequest={[Function]}
  modalButtonClassName="button"
  title="DELETE CENTER"
>
  <div
    className="delete-modal-content"
  >
    <p
      id="confirm-msg"
    >
      Are you sure you want to delete the 
      "Lagos"
       
      center? This cannot be undone & all data will be lost
    </p>
    <ActionButtons
      actionButtonText="DELETE CENTER"
      cancelButtonText="CANCEL"
      isLoading={false}
      onClickCancel={[Function]}
      onClickSubmit={[Function]}
      withCancel={true}
    />
  </div>
</MrmModal>
`;
